Migrate `mode` from TH to ATen

ysiraichi opened the pull request:
Fixes #24731 #24673 #24597 #24526 #46507
Related #24507

ysiraichi committed:
Port 'mode' (CPU) into ATen.

ysiraichi committed:
Port 'mode' (CUDA) to ATen.

ysiraichi committed:
Add scalar type checks.

ysiraichi committed:
Remove TH functions for 'mode'.

ysiraichi committed:
Add missing thrust includes.

vadimkantorov commented:
Would it also lead to better performance on unfold'ed tensors? https://github.com/pytorch/pytorch/issues/45009

ysiraichi committed:
Revert `_mode*` deletion.

ysiraichi commented:
@vadimkantorov I don't think so. The code is basically the same. There shouldn't be much change.

ysiraichi committed:
Use `std::sort` at every slice.

ysiraichi commented:
After benchmarking the changes, I noticed some regression on the CPU version. The only difference between the two versions were when to sort the input:
 
- The TH version sorted each "dim size" slice using a [quicksort implementation][1];
- My version sorted every slice at once in the beginning (using `at::sort`).

Turns out my approach got a slowdown of 0.82x. So, I rewrote parts of my implementation, using `std::sort`, which got comparable results to the original implementation.

 Find the experiment I ran below. In the CPU results:
- **v1**: the version using `at::sort` in the beginning;
- **v2**: the version using `std::sort` for each slice.
 
 [1]: https://github.com/pytorch/pytorch/blob/7763c127cd5630ba4123ad89fc5243c28e91aa4a/aten/src/TH/generic/THTensorMoreMath.cpp#L346
 
 <details>
 <summary>Benchmark Program</summary>
 
 ```python
import torch
import timeit
import sys

device = sys.argv[1]

sizes = [ 200, 500, 1000 ]
ts = [ 10000 ]
dims = 2
dtypes = [
    torch.uint8,
    torch.short,
    torch.int,
    torch.int64,
    torch.half,
    torch.float,
    torch.double
]

for sz in sizes:
        shape = (sz, sz)
        a = f"torch.randn({shape}, device = '{device}')"
        for dim in range(dims):
            for t in ts:
                print(f"torch.mode(a, {dim}) with size {shape} for {t} times {dtype}")
                print(timeit.timeit(
                    f"torch.mode(a, {dim}); torch.cuda.synchronize()",
                    setup = f"import torch; a = {a}",
                    number = t
                ))
 ```
 </details>
 
 <details>
 <summary>CPU Benchmark Results</summary>

  | Calls | Shape        | dim | dtype   | (TH) Time (s) | (v1) Time (s) | (v2) Time (s) |
  |-------|--------------|-----|---------|-----------|---------------|---------------|
  |  1000 | (200, 200)   |   0 | uint8   |      1.42 |          1.97 |          1.43 |
  |  1000 | (200, 200)   |   1 | uint8   |      1.43 |          1.96 |          1.39 |
  |  1000 | (200, 200)   |   0 | int16   |      1.41 |          1.98 |          1.41 |
  |  1000 | (200, 200)   |   1 | int16   |      1.42 |          2.00 |          1.39 |
  |  1000 | (200, 200)   |   0 | int32   |      1.42 |          2.00 |          1.40 |
  |  1000 | (200, 200)   |   1 | int32   |      1.40 |          1.94 |          1.39 |
  |  1000 | (200, 200)   |   0 | int64   |      1.42 |          2.03 |          1.39 |
  |  1000 | (200, 200)   |   1 | int64   |      1.42 |          1.95 |          1.39 |
  |  1000 | (200, 200)   |   0 | float32 |      1.42 |          1.98 |          1.38 |
  |  1000 | (200, 200)   |   1 | float32 |      1.42 |          1.96 |          1.41 |
  |  1000 | (200, 200)   |   0 | float64 |      1.43 |          2.01 |          1.42 |
  |  1000 | (200, 200)   |   1 | float64 |      1.42 |          1.99 |          1.43 |
  |  1000 | (500, 500)   |   0 | uint8   |     10.29 |         13.18 |         10.23 |
  |  1000 | (500, 500)   |   1 | uint8   |     10.27 |         12.17 |         10.07 |
  |  1000 | (500, 500)   |   0 | int16   |     10.33 |         13.11 |         10.20 |
  |  1000 | (500, 500)   |   1 | int16   |     10.27 |         12.16 |         10.07 |
  |  1000 | (500, 500)   |   0 | int32   |     10.30 |         13.02 |         10.21 |
  |  1000 | (500, 500)   |   1 | int32   |     10.37 |         12.04 |         10.01 |
  |  1000 | (500, 500)   |   0 | int64   |     10.34 |         12.76 |         10.15 |
  |  1000 | (500, 500)   |   1 | int64   |     10.25 |         12.09 |         10.07 |
  |  1000 | (500, 500)   |   0 | float32 |     10.27 |         12.82 |         10.18 |
  |  1000 | (500, 500)   |   1 | float32 |     10.25 |         12.02 |         10.07 |
  |  1000 | (500, 500)   |   0 | float64 |     10.33 |         12.67 |         10.16 |
  |  1000 | (500, 500)   |   1 | float64 |     10.39 |         12.10 |         10.06 |
  |  1000 | (1000, 1000) |   0 | uint8   |     45.64 |         55.97 |         44.98 |
  |  1000 | (1000, 1000) |   1 | uint8   |     45.51 |         49.68 |         44.41 |
  |  1000 | (1000, 1000) |   0 | int16   |     45.69 |         55.72 |         44.97 |
  |  1000 | (1000, 1000) |   1 | int16   |     45.52 |         49.76 |         44.25 |
  |  1000 | (1000, 1000) |   0 | int32   |     45.85 |         55.92 |         45.06 |
  |  1000 | (1000, 1000) |   1 | int32   |     45.48 |         49.58 |         44.37 |
  |  1000 | (1000, 1000) |   0 | int64   |     45.94 |         55.77 |         45.10 |
  |  1000 | (1000, 1000) |   1 | int64   |     45.57 |         49.88 |         44.01 |
  |  1000 | (1000, 1000) |   0 | float32 |     45.61 |         55.29 |         44.41 |
  |  1000 | (1000, 1000) |   1 | float32 |     45.28 |         49.73 |         44.10 |
  |  1000 | (1000, 1000) |   0 | float64 |     45.37 |         55.48 |         44.44 |
  |  1000 | (1000, 1000) |   1 | float64 |     45.10 |         50.04 |         43.69 |
</details>

<details>
 <summary>GPU Benchmark Results</summary>

  | Calls | Shape        | dim | dtype   | (TH) Time (s) | (ATen) Time (s) |
  |-------|--------------|-----|---------|-----------|-------------|
  | 10000 | (200, 200)   |   0 | uint8   |      0.80 |        0.81 |
  | 10000 | (200, 200)   |   1 | uint8   |      0.65 |        0.63 |
  | 10000 | (200, 200)   |   0 | int16   |      0.76 |        0.74 |
  | 10000 | (200, 200)   |   1 | int16   |      0.65 |        0.64 |
  | 10000 | (200, 200)   |   0 | int32   |      0.76 |        0.75 |
  | 10000 | (200, 200)   |   1 | int32   |      0.65 |        0.63 |
  | 10000 | (200, 200)   |   0 | int64   |      0.76 |        0.75 |
  | 10000 | (200, 200)   |   1 | int64   |      0.65 |        0.64 |
  | 10000 | (200, 200)   |   0 | float16 |      0.76 |        0.75 |
  | 10000 | (200, 200)   |   1 | float16 |      0.65 |        0.63 |
  | 10000 | (200, 200)   |   0 | float32 |      0.76 |        0.73 |
  | 10000 | (200, 200)   |   1 | float32 |      0.65 |        0.62 |
  | 10000 | (200, 200)   |   0 | float64 |      0.76 |        0.74 |
  | 10000 | (200, 200)   |   1 | float64 |      0.65 |        0.63 |
  | 10000 | (500, 500)   |   0 | uint8   |      1.36 |        1.34 |
  | 10000 | (500, 500)   |   1 | uint8   |      1.25 |        1.22 |
  | 10000 | (500, 500)   |   0 | int16   |      1.35 |        1.34 |
  | 10000 | (500, 500)   |   1 | int16   |      1.24 |        1.22 |
  | 10000 | (500, 500)   |   0 | int32   |      1.35 |        1.33 |
  | 10000 | (500, 500)   |   1 | int32   |      1.24 |        1.22 |
  | 10000 | (500, 500)   |   0 | int64   |      1.35 |        1.34 |
  | 10000 | (500, 500)   |   1 | int64   |      1.24 |        1.22 |
  | 10000 | (500, 500)   |   0 | float16 |      1.36 |        1.36 |
  | 10000 | (500, 500)   |   1 | float16 |      1.26 |        1.23 |
  | 10000 | (500, 500)   |   0 | float32 |      1.36 |        1.36 |
  | 10000 | (500, 500)   |   1 | float32 |      1.26 |        1.24 |
  | 10000 | (500, 500)   |   0 | float64 |      1.36 |        1.35 |
  | 10000 | (500, 500)   |   1 | float64 |      1.26 |        1.23 |
  | 10000 | (1000, 1000) |   0 | uint8   |      2.74 |        2.73 |
  | 10000 | (1000, 1000) |   1 | uint8   |      2.46 |        2.43 |
  | 10000 | (1000, 1000) |   0 | int16   |      2.74 |        2.74 |
  | 10000 | (1000, 1000) |   1 | int16   |      2.46 |        2.45 |
  | 10000 | (1000, 1000) |   0 | int32   |      2.76 |        2.74 |
  | 10000 | (1000, 1000) |   1 | int32   |      2.47 |        2.44 |
  | 10000 | (1000, 1000) |   0 | int64   |      2.76 |        2.74 |
  | 10000 | (1000, 1000) |   1 | int64   |      2.48 |        2.45 |
  | 10000 | (1000, 1000) |   0 | float16 |      2.76 |        2.74 |
  | 10000 | (1000, 1000) |   1 | float16 |      2.47 |        2.45 |
  | 10000 | (1000, 1000) |   0 | float32 |      2.76 |        2.77 |
  | 10000 | (1000, 1000) |   1 | float32 |      2.47 |        2.46 |
  | 10000 | (1000, 1000) |   0 | float64 |      2.76 |        2.77 |
  | 10000 | (1000, 1000) |   1 | float64 |      2.47 |        2.46 |
</details>

ysiraichi commented:
As far as I understand, the CI test errors do not seem related to this PR.

1. _pytorch_linux_xenial_py3_clang7_onnx_ort_test2_: ```ssl.py:631: ConnectionResetError```
2. _caffe2-pytorch-linux-bionic-rocm3.10-py3.6-test_: ```hypothesis.errors.Flaky: Hypothesis test_weighted_sample```

(1) feels like it's something CI related. Plus, (2) apparently calls a ```WeightedSum``` operator (probably defined [here][1]) that does not really use ```mode```.

[1]: https://github.com/pytorch/pytorch/blob/master/caffe2/operators/weighted_sample_op.cc

peterbell10 commented on aten/src/ATen/native/cuda/TensorModeKernel.cu:
```diff
@@ -0,0 +1,370 @@
+#if CUDA_VERSION >= 7000 || defined __HIP_PLATFORM_HCC__
+  auto stream = at::cuda::getCurrentCUDAStream();
+#endif
+  // THAssert(THCTensor_(isContiguous)(state, input));
```
```suggestion
  TORCH_INTERNAL_ASSERT(self.is_contiguous());
```

peterbell10 commented on aten/src/ATen/native/cpu/TensorCompareKernel.cpp:
```diff
@@ -213,6 +215,58 @@ static void isneginf_kernel_impl(TensorIterator& iter) {
+          int64_t temp_freq = 0;
+          int64_t max_freq = 0;
+
+          std::vector<std::pair<scalar_t, int64_t>> elements(self_dim_size);
+          std::vector<int64_t> idxs(self_dim_size);
```
This will allocate new memory `self.numel() / self.size(dim)` times whereas you only need one to allocate once per thread. I suggest adding a `compare_base_kernel` variant that lets the caller define their own `for_each` loop (line 50).

i.e. 
```cpp
static inline void compare_base_kernel_core(Tensor& result1, Tensor& result2, const Tensor& self, int64_t dim, bool keepdim, const TensorIterator::loop_t& loop);

template <typename scalar_t, typename scalar_t_2=int64_t, typename func_t>
static inline void compare_base_kernel(Tensor& result1, Tensor& result2,
    const Tensor& self,
    int64_t dim,
    bool keepdim,
    const func_t& f) {
  auto loop = ...; // From line 50
  return compare_base_kernel_core(result1, result2, self, dim, keepdim, loop);
}
```

Then you can call `compare_base_kernel_core` from this kernel and allocate the memory only once inside the loop.

ysiraichi replied:
That makes sense.
Thanks for the suggestion.

peterbell10 commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -285,18 +286,33 @@ Tensor _s_where(const Tensor& condition, const Tensor& self, const Tensor& other
 }
 
+std::tuple<Tensor, Tensor> _mode(const Tensor& self, int64_t dim, bool keepdim) {
+  return at::native::mode(self, dim, keepdim);
```
You can just drop the `_mode` operator. Since it starts with an underscore, users shouldn't be relying on it.  Just add an exception to the backward compatibility check:
https://github.com/pytorch/pytorch/blob/d22f700f9ea566f07ad392b40889aae1697303a6/test/backward_compatibility/check_backward_compatibility.py#L75

peterbell10 reviewed (changes requested):
In this porting PR, you should also remove the code in `TH` and `THC` which this replaces.

peterbell10 commented on aten/src/ATen/native/cuda/TensorModeKernel.cu:
```diff
@@ -0,0 +1,370 @@
+#if defined(THC_REAL_IS_HALF)
+          ,
+      ThrustHalfLess()
+#endif
```
Or hey, it's compiling already so might not need a comparator at all.

ysiraichi replied:
Got it. I removed all of them.

ysiraichi commented on aten/src/ATen/native/cuda/TensorModeKernel.cu:
```diff
@@ -0,0 +1,370 @@
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    THCudaLongStorage* sort_buffer,
+    THLongStorage* position,
```
Got it.
So, I guess `position` can be a `std::vector` (since it's only used in the host), right!?

By the way, there are a bunch of functions in [TensorModeKernel.cuh][1] that use `TH` functions. Should I also get rid of all of them (or bring them to ATen)? For example, it uses:

- `reduceBlockWithNThreadLocalReductions` from [THCReduceApplyUtils.cuh][2];
- `inclusivePrefixScan` from [THCScanUtils.cuh][3];
- `bitonicSortKeys` from [THCSortUtils.cuh][4].

[1]: https://github.com/pytorch/pytorch/pull/52043/files#diff-8327a1711c8c3c8da49165a4d43d5d9a9c13ad521ebac7e7dbda2ee6ce36e076
[2]: https://github.com/pytorch/pytorch/blob/1903b32c35a7e7432d62e43a2d3a679f6f01066d/aten/src/THC/THCReduceApplyUtils.cuh#L128
[3]: https://github.com/pytorch/pytorch/blob/1903b32c35a7e7432d62e43a2d3a679f6f01066d/aten/src/THC/THCScanUtils.cuh#L32
[4]: https://github.com/pytorch/pytorch/blob/1903b32c35a7e7432d62e43a2d3a679f6f01066d/aten/src/THC/THCSortUtils.cuh#L104

peterbell10 replied:
>So, I guess position can be a std::vector (since it's only used in the host), right!?

Yeah, that's fine.

>By the way, there are a bunch of functions in TensorModeKernel.cuh that use TH functions. Should I also get rid of all of them (or bring them to ATen)? 

I'd say they should be moved to ATen. A `git grep` shows these functions are only used by mode.

ysiraichi replied:
> I'd say they should be moved to ATen. A git grep shows these functions are only used by mode.

I moved those functions to the header file.
But, after doing so, the previous `mode` implementation broke (since I moved the code).
Therefore, I thought that removing that implementation completely was the way to go.
Is that ok?

That said, I haven't removed the old `mode` implementation for CPUs. Should I do that?

peterbell10 replied:
Yes, you should remove all the old TH/THC code for `mode`.

ysiraichi replied:
Got it.

ysiraichi committed:
Implement suggestions.

- remove `_mode` tensor operation;
- [CPU] hoist vector allocation (and initialization) out of the inner
    loop;
- [GPU] replace few calls to TH code with ATen.

ysiraichi committed:
Fix regression on smaller types.

ysiraichi committed:
Include less TH libs.

ysiraichi committed:
Move code that only `mode` touches to ATen.

ysiraichi committed:
Remove `mode` CUDA TH implementation.

ysiraichi commented:
I noticed the benchmark program I posted in my [previous comment][1] was wrong.
Turns out I forgot to change the `dtype` of the tensor, before applying the mode.
After re-running the tests, the numbers make much more sense, now.
So, **the only difference** is that I am applying `mode` to the correct `dtype`.
Below, I am showing the following versions:

- **TH:** the base version (TH);
- **v2:** the previous version we had;
- **v3:** the version after implementing the suggestions by @peterbell10 plus
    replacing the comparison function `std::sort` uses by one that only
    considers the first element in the pair (biggest improvement).

[1]: https://github.com/pytorch/pytorch/pull/52043#issuecomment-777434604

<details>
<summary>Benchmark Program</summary>

 ```python
import torch
import timeit
import sys

device = sys.argv[1]

sizes = [ 200, 500, 1000 ]

## 1000 for CPUs
ts = [ 10000 ]

dims = 2
dtypes = [
    torch.uint8,
    torch.short,
    torch.int,
    torch.int64,
    torch.half,
    torch.float,
    torch.double
]

## Warming up!
timeit.timeit(
    f"torch.mode(a, 0); torch.cuda.synchronize()",
    setup = f"import torch; a = torch.randn((200, 200), device = '{device}'); a = a.to(torch.uint8)",
    number = 1000
)

for sz in sizes:
        shape = (sz, sz)
        a = f"torch.randn({shape}, device = '{device}')"
        for dim in range(dims):
            for t in ts:
                print(f"torch.mode(a, {dim}) with size {shape} for {t} times {dtype}")
                print(timeit.timeit(
                    f"torch.mode(a, {dim}); torch.cuda.synchronize()",
                    setup = f"import torch; a = {a}; a = a.to({dtype})",
                    number = t
                ))
 ```
 </details>
 
<details>
<summary>CPU Benchmark Results</summary>

| Calls | Shape        | dim | dtype   | (TH) Time | (v2) Time (s) | (v3) Time (s) |
|-------|--------------|-----|---------|-----------|---------------|---------------|
|  1000 | (200, 200)   |   0 | uint8   |      0.47 |          1.22 |          0.47 |
|  1000 | (200, 200)   |   1 | uint8   |      0.47 |          1.19 |          0.47 |
|  1000 | (200, 200)   |   0 | int16   |      0.48 |          1.23 |          0.46 |
|  1000 | (200, 200)   |   1 | int16   |      0.48 |          1.21 |          0.47 |
|  1000 | (200, 200)   |   0 | int32   |      0.48 |          1.24 |          0.47 |
|  1000 | (200, 200)   |   1 | int32   |      0.48 |          1.22 |          0.47 |
|  1000 | (200, 200)   |   0 | int64   |      0.49 |          1.21 |          0.48 |
|  1000 | (200, 200)   |   1 | int64   |      0.48 |          1.24 |          0.47 |
|  1000 | (200, 200)   |   0 | float32 |      1.43 |          1.42 |          1.32 |
|  1000 | (200, 200)   |   1 | float32 |      1.41 |          1.43 |          1.31 |
|  1000 | (200, 200)   |   0 | float64 |      1.44 |          1.43 |          1.32 |
|  1000 | (200, 200)   |   1 | float64 |      1.42 |          1.42 |          1.33 |
|  1000 | (500, 500)   |   0 | uint8   |      3.04 |          8.48 |          3.16 |
|  1000 | (500, 500)   |   1 | uint8   |      3.02 |          8.25 |          3.05 |
|  1000 | (500, 500)   |   0 | int16   |      3.12 |          8.48 |          2.99 |
|  1000 | (500, 500)   |   1 | int16   |      3.09 |          8.35 |          2.94 |
|  1000 | (500, 500)   |   0 | int32   |      3.15 |          8.56 |          3.03 |
|  1000 | (500, 500)   |   1 | int32   |      3.10 |          8.46 |          2.97 |
|  1000 | (500, 500)   |   0 | int64   |      3.25 |          8.39 |          3.13 |
|  1000 | (500, 500)   |   1 | int64   |      3.16 |          8.29 |          3.08 |
|  1000 | (500, 500)   |   0 | float32 |     10.39 |         10.18 |          9.59 |
|  1000 | (500, 500)   |   1 | float32 |     10.33 |         10.13 |          9.59 |
|  1000 | (500, 500)   |   0 | float64 |     10.91 |         10.15 |          9.61 |
|  1000 | (500, 500)   |   1 | float64 |     10.89 |         10.15 |          9.60 |
|  1000 | (1000, 1000) |   0 | uint8   |     12.63 |         35.56 |         12.95 |
|  1000 | (1000, 1000) |   1 | uint8   |     12.61 |         35.48 |         12.61 |
|  1000 | (1000, 1000) |   0 | int16   |     13.13 |         36.18 |         12.15 |
|  1000 | (1000, 1000) |   1 | int16   |     12.79 |         35.74 |         11.91 |
|  1000 | (1000, 1000) |   0 | int32   |     12.96 |         37.00 |         12.38 |
|  1000 | (1000, 1000) |   1 | int32   |     12.65 |         36.56 |         12.09 |
|  1000 | (1000, 1000) |   0 | int64   |     13.35 |         36.17 |         13.01 |
|  1000 | (1000, 1000) |   1 | int64   |     13.00 |         35.87 |         12.65 |
|  1000 | (1000, 1000) |   0 | float32 |     45.33 |         44.73 |         41.75 |
|  1000 | (1000, 1000) |   1 | float32 |     45.29 |         44.27 |         41.54 |
|  1000 | (1000, 1000) |   0 | float64 |     46.82 |         45.17 |         42.11 |
|  1000 | (1000, 1000) |   1 | float64 |     46.83 |         45.19 |         41.89 |
</details>

<details>
<summary>GPU Benchmark Results</summary>

| Calls | Shape        | dim | dtype   | (TH) Time | (ATen) Time |
|-------|--------------|-----|---------|-----------|-------------|
| 10000 | (200, 200)   |   0 | uint8   |      0.76 |        0.76 |
| 10000 | (200, 200)   |   1 | uint8   |      0.61 |        0.60 |
| 10000 | (200, 200)   |   0 | int16   |      0.71 |        0.72 |
| 10000 | (200, 200)   |   1 | int16   |      0.61 |        0.60 |
| 10000 | (200, 200)   |   0 | int32   |      0.78 |        0.75 |
| 10000 | (200, 200)   |   1 | int32   |      0.65 |        0.64 |
| 10000 | (200, 200)   |   0 | int64   |      0.86 |        0.87 |
| 10000 | (200, 200)   |   1 | int64   |      0.75 |        0.74 |
| 10000 | (200, 200)   |   0 | float16 |      0.73 |        0.72 |
| 10000 | (200, 200)   |   1 | float16 |      0.61 |        0.60 |
| 10000 | (200, 200)   |   0 | float32 |      0.75 |        0.74 |
| 10000 | (200, 200)   |   1 | float32 |      0.64 |        0.63 |
| 10000 | (200, 200)   |   0 | float64 |      0.90 |        0.91 |
| 10000 | (200, 200)   |   1 | float64 |      0.79 |        0.80 |
| 10000 | (500, 500)   |   0 | uint8   |      1.27 |        1.29 |
| 10000 | (500, 500)   |   1 | uint8   |      1.16 |        1.17 |
| 10000 | (500, 500)   |   0 | int16   |      1.27 |        1.28 |
| 10000 | (500, 500)   |   1 | int16   |      1.17 |        1.17 |
| 10000 | (500, 500)   |   0 | int32   |      1.38 |        1.38 |
| 10000 | (500, 500)   |   1 | int32   |      1.27 |        1.27 |
| 10000 | (500, 500)   |   0 | int64   |      1.66 |        1.67 |
| 10000 | (500, 500)   |   1 | int64   |      1.54 |        1.54 |
| 10000 | (500, 500)   |   0 | float16 |      1.28 |        1.29 |
| 10000 | (500, 500)   |   1 | float16 |      1.17 |        1.18 |
| 10000 | (500, 500)   |   0 | float32 |      1.34 |        1.34 |
| 10000 | (500, 500)   |   1 | float32 |      1.23 |        1.23 |
| 10000 | (500, 500)   |   0 | float64 |      1.73 |        1.74 |
| 10000 | (500, 500)   |   1 | float64 |      1.61 |        1.62 |
| 10000 | (1000, 1000) |   0 | uint8   |      2.40 |        2.41 |
| 10000 | (1000, 1000) |   1 | uint8   |      2.19 |        2.19 |
| 10000 | (1000, 1000) |   0 | int16   |      2.45 |        2.45 |
| 10000 | (1000, 1000) |   1 | int16   |      2.22 |        2.22 |
| 10000 | (1000, 1000) |   0 | int32   |      2.73 |        2.73 |
| 10000 | (1000, 1000) |   1 | int32   |      2.44 |        2.43 |
| 10000 | (1000, 1000) |   0 | int64   |      3.38 |        3.39 |
| 10000 | (1000, 1000) |   1 | int64   |      3.04 |        3.03 |
| 10000 | (1000, 1000) |   0 | float16 |      2.50 |        2.49 |
| 10000 | (1000, 1000) |   1 | float16 |      2.27 |        2.26 |
| 10000 | (1000, 1000) |   0 | float32 |      2.74 |        2.74 |
| 10000 | (1000, 1000) |   1 | float32 |      2.45 |        2.45 |
| 10000 | (1000, 1000) |   0 | float64 |      3.59 |        3.61 |
| 10000 | (1000, 1000) |   1 | float64 |      3.24 |        3.24 |
</details>

ysiraichi commented:
I noticed that the current (and previous) implementation of  `mode` is not being executed in parallel for any of the tensors in the benchmark I created. Apparently, that's because of [this condition][1] in the implementation of a `TensorIterator`.

The problem appears to be that it checks whether there are, at least, `internal::GRAIN_SIZE` (32768) elements to be calculated. Since the tensor size is 1000x1000, there are only 1000 resulting mode elements, i.e. it won't be parallelized. Is that the intended behavior?

[1]: https://github.com/pytorch/pytorch/blob/8c185e62f99c272bf991b8e0558fccb5f8c780c0/aten/src/ATen/TensorIterator.cpp#L685

peterbell10 commented:
I think that condition should be using the `grain_size` argument, not `internal::GRAIN_SIZE`. Looks like an oversight from when `grain_size` was added (#29705).

ysiraichi commented:
All the GPU results I showed until now correspond to the fused kernel. In this post, I benchmarked the non-fused kernel (which is orders of magnitude slower). The benchmark program is basically the same. Note that all tensors are 2D, where the smallest dimension has size 200 (since it just iterates in the host side).

<details>
    <summary>Non-Fused GPU Results</summary>

| Calls | Dim Shape | dim | dtype         | (TH) Time | (ATen) Time |
|-------|-----------|-----|---------------|-----------|-------------|
|   100 |      2000 |   0 | torch.uint8   |      7.94 |        4.00 |
|   100 |      2000 |   1 | torch.uint8   |      7.75 |        3.89 |
|   100 |      2000 |   0 | torch.int16   |      7.86 |        3.97 |
|   100 |      2000 |   1 | torch.int16   |      7.80 |        3.97 |
|   100 |      2000 |   0 | torch.int32   |      8.12 |        4.09 |
|   100 |      2000 |   1 | torch.int32   |      8.13 |        4.10 |
|   100 |      2000 |   0 | torch.int64   |      8.26 |        4.40 |
|   100 |      2000 |   1 | torch.int64   |      8.21 |        4.41 |
|   100 |      2000 |   0 | torch.float16 |      8.32 |        4.42 |
|   100 |      2000 |   1 | torch.float16 |      8.27 |        4.49 |
|   100 |      2000 |   0 | torch.float32 |      8.21 |        4.48 |
|   100 |      2000 |   1 | torch.float32 |      8.10 |        4.44 |
|   100 |      2000 |   0 | torch.float64 |      8.59 |        4.52 |
|   100 |      2000 |   1 | torch.float64 |      8.57 |        4.49 |
|   100 |      4000 |   0 | torch.uint8   |      8.46 |        4.32 |
|   100 |      4000 |   1 | torch.uint8   |      8.59 |        4.25 |
|   100 |      4000 |   0 | torch.int16   |      9.29 |        5.04 |
|   100 |      4000 |   1 | torch.int16   |      9.18 |        5.03 |
|   100 |      4000 |   0 | torch.int32   |      9.88 |        6.04 |
|   100 |      4000 |   1 | torch.int32   |      9.89 |        6.08 |
|   100 |      4000 |   0 | torch.int64   |     12.17 |        8.34 |
|   100 |      4000 |   1 | torch.int64   |     12.20 |        8.35 |
|   100 |      4000 |   0 | torch.float16 |      8.51 |        4.12 |
|   100 |      4000 |   1 | torch.float16 |      8.45 |        4.32 |
|   100 |      4000 |   0 | torch.float32 |     10.17 |        6.24 |
|   100 |      4000 |   1 | torch.float32 |     10.11 |        6.21 |
|   100 |      4000 |   0 | torch.float64 |     12.52 |        8.68 |
|   100 |      4000 |   1 | torch.float64 |     12.62 |        8.67 |
|   100 |      8000 |   0 | torch.uint8   |      8.31 |        4.10 |
|   100 |      8000 |   1 | torch.uint8   |      8.51 |        4.20 |
|   100 |      8000 |   0 | torch.int16   |      9.36 |        5.09 |
|   100 |      8000 |   1 | torch.int16   |      9.22 |        5.07 |
|   100 |      8000 |   0 | torch.int32   |     10.01 |        6.12 |
|   100 |      8000 |   1 | torch.int32   |      9.98 |        6.12 |
|   100 |      8000 |   0 | torch.int64   |     12.41 |        8.53 |
|   100 |      8000 |   1 | torch.int64   |     12.42 |        8.53 |
|   100 |      8000 |   0 | torch.float16 |      8.53 |        4.26 |
|   100 |      8000 |   1 | torch.float16 |      8.75 |        4.32 |
|   100 |      8000 |   0 | torch.float32 |     10.33 |        6.35 |
|   100 |      8000 |   1 | torch.float32 |     10.34 |        6.38 |
|   100 |      8000 |   0 | torch.float64 |     12.72 |        8.93 |
|   100 |      8000 |   1 | torch.float64 |     12.86 |        8.94 |
</details>

ysiraichi committed:
Remove `mode` CPU TH implementation.

ysiraichi committed:
Update comments.

ysiraichi committed:
Replace macro for templated function.

ysiraichi commented:
> I think that condition should be using the `grain_size` argument, not `internal::GRAIN_SIZE`. Looks like an oversight from when `grain_size` was added (#29705).

Should I leave it to a future PR?

peterbell10 commented on aten/src/ATen/native/cpu/TensorCompareKernel.cpp:
```diff
@@ -213,6 +229,75 @@ static void isneginf_kernel_impl(TensorIterator& iter) {
+        int64_t temp_freq = 0;
+        int64_t max_freq = 0;
+
+        std::transform(
+            idxs.begin(), idxs.end(), elements.begin(), [=](int64_t i) {
+                return std::make_pair(self_data[i * self_dim_stride], i);
+            });
```
Am I right in saying `idxs` is never modified? In that case, I'd just make it a loop and not use transform. That way you don't need to allocate any vector.

ysiraichi replied:
Yes, `idxs` won't change. But, won't `transform` just use the allocated space (instead of reallocating it)?

peterbell10 replied:
Sure, it's not a huge performance issue. Just why allocate that memory at all?

ysiraichi replied:
Ah, right. I think I wasn't getting exactly what you were talking about. That does make sense, though! I will do that.

peterbell10 commented on aten/src/ATen/native/cuda/TensorModeKernel.cu:
```diff
@@ -0,0 +1,321 @@
+    indices_data += ensure_nonempty_stride(indices, i) * pos;
+  }
+
+  AT_CUDA_CHECK(cudaMemcpyAsync(
+      values_data, &mode, sizeof(scalar_t), cudaMemcpyHostToDevice, stream));
+  AT_CUDA_CHECK(cudaMemcpyAsync(
+      indices_data, &index, sizeof(scalar_t), cudaMemcpyHostToDevice, stream));
```
This is dangerous because the copy might not have finished by the time `mode` and `index` go out of scope. You'll need an `AT_CUDA_CHECK(cudaStreamSynchronize(stream))` after.

peterbell10 commented:
> Should I leave this for a future PR?

Submitting it as a seperate PR would make sense but it's also worth benchmarking with the fix in place.

ysiraichi committed:
Loopify `transform` call and sync CUDA for naive kernel.

ysiraichi commented:
Here are the benchmarks after a possible fix to `TensorIterator::for_each`. For comparison (and completeness), I am showing a few versions:

- **v3:** the last version from the previous benchmark results;
- **v4:** the version after replacing `transform` with a loop;
- **par:** the version after applying the fix (`num_threads = 24`).

**The fix:** as suggested in [this comment][1], replacing the grain size to be used, i.e. replace `internal::GRAIN_SIZE` with `grain_size`.

[1]: https://github.com/pytorch/pytorch/pull/52043#issuecomment-780607094

<details>
    <summary>Parallel CPU Results</summary>
    
| Calls | Shape        | dim | dtype   | (TH) Time | (v3) Time (s) | (v4) Time (s) | (par) Time (s) |
|-------|--------------|-----|---------|-----------|---------------|---------------|----------------|
|  1000 | (200, 200)   |   0 | uint8   |      0.47 |          0.47 |          0.45 |           0.06 |
|  1000 | (200, 200)   |   1 | uint8   |      0.47 |          0.47 |          0.46 |           0.06 |
|  1000 | (200, 200)   |   0 | int16   |      0.48 |          0.46 |          0.47 |           0.07 |
|  1000 | (200, 200)   |   1 | int16   |      0.48 |          0.47 |          0.47 |           0.07 |
|  1000 | (200, 200)   |   0 | int32   |      0.48 |          0.47 |          0.48 |           0.06 |
|  1000 | (200, 200)   |   1 | int32   |      0.48 |          0.47 |          0.47 |           0.06 |
|  1000 | (200, 200)   |   0 | int64   |      0.49 |          0.48 |          0.50 |           0.06 |
|  1000 | (200, 200)   |   1 | int64   |      0.48 |          0.47 |          0.51 |           0.06 |
|  1000 | (200, 200)   |   0 | float32 |      1.43 |          1.32 |          1.31 |           0.09 |
|  1000 | (200, 200)   |   1 | float32 |      1.41 |          1.31 |          1.30 |           0.10 |
|  1000 | (200, 200)   |   0 | float64 |      1.44 |          1.32 |          1.32 |           0.09 |
|  1000 | (200, 200)   |   1 | float64 |      1.42 |          1.33 |          1.32 |           0.09 |
|  1000 | (500, 500)   |   0 | uint8   |      3.04 |          3.16 |          3.09 |           0.25 |
|  1000 | (500, 500)   |   1 | uint8   |      3.02 |          3.05 |          3.04 |           0.24 |
|  1000 | (500, 500)   |   0 | int16   |      3.12 |          2.99 |          2.99 |           0.25 |
|  1000 | (500, 500)   |   1 | int16   |      3.09 |          2.94 |          2.92 |           0.25 |
|  1000 | (500, 500)   |   0 | int32   |      3.15 |          3.03 |          3.13 |           0.23 |
|  1000 | (500, 500)   |   1 | int32   |      3.10 |          2.97 |          3.12 |           0.22 |
|  1000 | (500, 500)   |   0 | int64   |      3.25 |          3.13 |          3.14 |           0.24 |
|  1000 | (500, 500)   |   1 | int64   |      3.16 |          3.08 |          3.13 |           0.24 |
|  1000 | (500, 500)   |   0 | float32 |     10.39 |          9.59 |          9.51 |           0.54 |
|  1000 | (500, 500)   |   1 | float32 |     10.33 |          9.59 |          9.45 |           0.56 |
|  1000 | (500, 500)   |   0 | float64 |     10.91 |          9.61 |          9.54 |           0.57 |
|  1000 | (500, 500)   |   1 | float64 |     10.89 |          9.60 |          9.52 |           0.51 |
|  1000 | (1000, 1000) |   0 | uint8   |     12.63 |         12.95 |         12.72 |           0.95 |
|  1000 | (1000, 1000) |   1 | uint8   |     12.61 |         12.61 |         12.72 |           0.93 |
|  1000 | (1000, 1000) |   0 | int16   |     13.13 |         12.15 |         12.44 |           0.85 |
|  1000 | (1000, 1000) |   1 | int16   |     12.79 |         11.91 |         12.17 |           1.00 |
|  1000 | (1000, 1000) |   0 | int32   |     12.96 |         12.38 |         13.31 |           0.99 |
|  1000 | (1000, 1000) |   1 | int32   |     12.65 |         12.09 |         13.12 |           0.80 |
|  1000 | (1000, 1000) |   0 | int64   |     13.35 |         13.01 |         13.33 |           0.78 |
|  1000 | (1000, 1000) |   1 | int64   |     13.00 |         12.65 |         13.17 |           0.80 |
|  1000 | (1000, 1000) |   0 | float32 |     45.33 |         41.75 |         42.30 |           3.07 |
|  1000 | (1000, 1000) |   1 | float32 |     45.29 |         41.54 |         42.06 |           3.08 |
|  1000 | (1000, 1000) |   0 | float64 |     46.82 |         42.11 |         42.50 |           3.12 |
|  1000 | (1000, 1000) |   1 | float64 |     46.83 |         41.89 |         42.34 |           3.07 |
</details>

peterbell10 commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -73,6 +73,7 @@
     ("aten::_foreach_div_", datetime.date(2021, 2, 25)),
     ("aten::_foreach_addcdiv", datetime.date(2021, 2, 25)),
     ("aten::mkldnn_linear", datetime.date(2021, 3, 2)),
+    ("aten::_mode", datetime.date(9999, 1, 1)),
```
When choosing a date here, it should be just enough time to merge the changes. Not absurdly far in the future. I usually go a month or so ahead to prevent issues.

ysiraichi replied:
Got it.

peterbell10 reviewed (commented):
Looks great, just a small comment.

ysiraichi committed:
Change backward compatibility `allow_list` date for `_mode`.

ngimel commented on aten/src/ATen/native/cuda/TensorModeKernel.cu:
```diff
@@ -0,0 +1,322 @@
+    // maximum number of blocks for a kernel launch
+    // 3. Can use 32-bit index math for indexing (mainly just for implementation
+    // conciseness, could be changed)
+    if (slice_size <= MAX_BLOCK_SIZE && slices <= MAX_GRID_SIZE &&
```
slices <= MAX_GRID_SIZE looks too restrictive, `getGridFromTiles` requires slices <= MAX_GRID_SIZE * MAX_GRID_SIZE * MAX_GRID_SIZE`. Relaxing this would probably fix #46507.
Can you also add a test for > 65535 slices?

ysiraichi replied:
Ah, I hadn't noticed that. That makes sense, indeed.
I believe it makes sense to rewrite it as `slices <= MAX_GRID_SIZE * MAX_GRID_SIZE * MAX_GRID_SIZE`, right!?

ngimel replied:
Yes, that should be the correct condition

ngimel commented on aten/src/ATen/native/cuda/TensorModeKernel.cu:
```diff
@@ -0,0 +1,322 @@
+    }
+  }
+
+  values.resize_(self_sizes);
```
this would result in the wrong behavior if values and indices provided by user are discontiguous yet have correct size, see expected behavior https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch

ysiraichi replied:
Do you mean that: in such a situation, `mode` would return contiguous `values` and `indices` instead of discontiguous values, right? Not sure I can see how that might happen, though... As far as I understand, given that the output tensors **have the correct size**, there are two cases:
1. `keepdims = True`: both output tensors have the same size as `self_sizes` (size 1 at `dim`)
2. `keepdims = False` (the scenario you are referring to): the condition `values.ndimension() >= dim` (same for `indices`) will always be true. Therefore, we will call `unsqueeze_` for them before `resize_`.

When `values` and `indices` have the correct size, `resize_` does nothing (AFAIK). Am I missing something? As it is now:

```python
> v.is_contiguous(), i.is_contiguous()
(False, False)
> torch.mode(a, dim = dim, keepdims = False, out = (v, i))
> v.is_contiguous(), i.is_contiguous()
(False, False)
```

peterbell10 replied:
I believe @ysiraichi is correct, although this should probably be using `resize_output` instead of `resize_`. Note that `resize_output` unconditionally calls `resize_(shape)` internally, so this is entirely about the warning message.

https://github.com/pytorch/pytorch/blob/d54be1a9467db5075256d229ea1b01f1a4bcba8d/aten/src/ATen/native/Resize.cpp#L22

Also note that this `resize_` call is copied from existing code, so both need to be changed to use `resize_output`:
https://github.com/pytorch/pytorch/blob/d54be1a9467db5075256d229ea1b01f1a4bcba8d/aten/src/ATen/native/cpu/TensorCompareKernel.cpp#L36-L37

ngimel replied:
Sorry, I didn't notice `unsqueeze_` above, the behavior is actually correct. @peterbell10 is right, resize_output should be called here.

ngimel commented on aten/src/ATen/native/cuda/TensorModeKernel.cu:
```diff
@@ -0,0 +1,322 @@
+                   thrust::not_equal_to<scalar_t>());
+
+  // Count frequency of each element
+  thrust::device_vector<scalar_t> keys(unique);
```
thrust::device_vector calls cudaMalloc under the hood, please avoid using it (you can allocate raw memory with cuda allocator and wrap it in thrust::device_ptr). 
Alternatively, look here for how to get counts of sorted unique elements without device_vectors https://github.com/pytorch/pytorch/blob/758fb94fcb8e20d41df6b055e80725e37ddb4854/aten/src/ATen/native/cuda/Unique.cu#L60-L68

ysiraichi replied:
I tried to keep the same logic without using `thrust::device_vector`. Is that alright?

ngimel replied:
Yeah, if you are allocating memory using pytorch allocator and not `thrust::device_vector` that should be fine.

peterbell10 commented:
It used quick sort before which isn't stable either. The documentation also doesn't make such a strong claim; it just says "indices is the index location of each mode value found".

ysiraichi committed:
Add checks for output device and dtypes.

ysiraichi committed:
Relax conditions for using fused kernel.

- Stop using `thrust::device_vector` (using `Tensor` instead).
- Allocating `sort_buffer` for every slice.
- Fused kernel conditions:
    - Slice size can be up to 2048 elements
    - Number of slices can be up to MAX_GRID_SIZE^3

ysiraichi committed:
Create tests for `mode` operator.

- Testing for wrong output dtype and device.
- Testing for large tensors that:
    - now use the fused kernel (but used the naive one before)
    - must use the naive kernel

ysiraichi committed:
Use `resize_output` instead of `resize_`.

ngimel commented:
@ysiraichi speaking of tests, can you please add OpInfo tests for mode? Usually when porting operations we also try to add them to OpInfo. Thanks!

mruberry commented:
> @ysiraichi speaking of tests, can you please add OpInfo tests for mode? Usually when porting operations we also try to add them to OpInfo. Thanks!

cc @heitorschueroff

ysiraichi committed:
Add OpInfo tests.

ngimel commented on test/test_reductions.py:
```diff
@@ -461,6 +461,82 @@ def test_mode(self, device):
+        values, indices = torch.mode(x, -1, False)
+
+        # Check whether the indices that correspond to each mode found are within
+        # one of the defined intervals
```
Just checking here that
```
(x.gather(1, indices.unsqueeze(1)).t() == values).all() 
```
is enough and more readable.

ysiraichi replied:
That makes sense.

ngimel commented on test/test_reductions.py:
```diff
@@ -461,6 +461,82 @@ def test_mode(self, device):
+        for (beg, end) in intervals:
+            b = b + ((indices >= beg) * (indices < end))
+
+        self.assertEqual((values == v).sum(), values.numel())
```
self.assertTrue((values == v).all().item())

ngimel commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1991,6 +2007,14 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+           dtypesIfCPU=all_types(),
+           dtypesIfCUDA=all_types_and(torch.float16),
+           test_inplace_grad=False,
+           supports_tensor_out=False,
```
it's `supports_out` now

ysiraichi replied:
I fixed it! Thanks.

ngimel reviewed (commented):
This looks good now, I've left minor comments about the tests, there are also a fair number of merge conflicts that need to be resolved.

ysiraichi committed:
Resolve conflicts for merging.

- Add a `loop1d_t` type template to the cpu kernel
- Implement suggestions for checking correctness in tests
- Remove `mode` from `method_tests`

ysiraichi committed:
Fix flake errors.

ngimel commented:
Test errors are real, can you please add cuda bfloat16 support?

ysiraichi committed:
Add support for `bool` and `bfloat16`.

ysiraichi commented:
Sure! I added tests for both `bool`and `bfloat16`.
Not sure if the reason is the same for others (#48978), but `test_variant_consistency_jit_mode_cpu_bfloat16` seems to be failing because the `ScatterGather` kernels are not implemented for `bfloat16`. Is that a bug?

mruberry commented:
> Sure! I added support for both `bool`and `bfloat16`.
> Not sure if the reason is the same for others (#48978), but `test_variant_consistency_jit_mode_cpu_bfloat16` seems to be failing because the `ScatterGather` kernels are not implemented for `bfloat16`. Is that a bug (i.e. is `ScatterGather` supposed not to support these types for some reason)?

The variant consistency tests now only run in float32, but you'll need to rebase and retest this and see if the new test_supported_backward tests need a skip.

ysiraichi committed:
Fix tests after rebasing.

ngimel commented:
@ysiraichi are you working on this PR? Can you please rebase and make sure the tests pass?

ysiraichi commented:
Sure. I will rebase this branch.

ysiraichi committed:
Add support to float16 for CPUs.

ysiraichi committed:
Revert removed sort declaration.

ysiraichi committed:
Remove 'supports_inplace_autograd' from 'mode' OpInfo.

ysiraichi committed:
Extend backward compatibility date.

ysiraichi committed:
Set test for CPU and CUDA only.

ezyang commented:
ACKing native_functions.yaml

If you're interested in a follow up task on this, making it structured would be pretty useful: https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

ysiraichi commented:
@ezyang that sounds like a great idea. I didn't find any issue for making `mode` structured, though. Should I just go ahead and open a PR for that?

ngimel commented:
Sorry @ysiraichi, yesterday there were problems with diff landing, and today there's a merge conflict, can you please fix and rebase again?

agolynski commented:
This breaks backwards compatibility test, 

https://app.circleci.com/pipelines/github/pytorch/pytorch/295702/workflows/40e395dc-01d5-427e-a765-078c09706e2b/jobs/12098048

Apr 03 06:47:11 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not. 
Apr 03 06:47:11 
Apr 03 06:47:11 Broken ops: [
Apr 03 06:47:11 	aten::_mode.values(Tensor self, int dim=-1, bool keepdim=False, *, Tensor(a!) values, Tensor(b!) indices) -> (Tensor(a!), Tensor(b!))
Apr 03 06:47:11 	aten::_mode(Tensor self, int dim=-1, bool keepdim=False) -> (Tensor, Tensor)
Apr 03 06:47:11 ]

I wonder why aten::_mode entry was set to be valid until Apr 2 as it was outdated at the time when PR was merged?

ezyang commented:
The PR was opened Feb. Typically authors set a window of a few months, and this PR just took too long to get merged. Please forward fix.
